internal/notifier/notifier.go

```diff
@@ -220,122 +362,175 @@ func (n *Notifier) runOnce(ctx context.Context) error {
-	// automatically when the conn restarts.
-	for {
-		select {
-		case <-n.subChangeCh:
-		default:
-			return
-		}
+	// Notably, these unlock then lock again, the reverse of what you'd normally
+	// expect in a mutex pattern. This is because this function is only expected
+	// to be called with the mutex already locked, but we need to unlock it to
+	// give the main loop a chance to run interrupt operations.
+	n.mu.Unlock()
+	defer n.mu.Lock()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-time.After(5 * time.Second):
+		return errors.New("timed out waiting for interrupt operation")
 	}
 }
 
-func (n *Notifier) Listen(topic NotificationTopic, notifyFunc NotifyFunc) *Subscription {
+func (n *Notifier) Listen(ctx context.Context, topic NotificationTopic, notifyFunc NotifyFunc) (*Subscription, error) {
 	n.mu.Lock()
 	defer n.mu.Unlock()
 
 	sub := &Subscription{
-		creationTime: time.Now(),
-		topic:        topic,
-		notifyFunc:   notifyFunc,
-		unlistenOnce: &sync.Once{},
-		notifier:     n,
+		notifyFunc: notifyFunc,
+		topic:      topic,
+		notifier:   n,
 	}
 
-	isNewTopic := false
-
-	subs := n.subs[topic]
-	if subs == nil {
-		isNewTopic = true
-		subs = make([]*Subscription, 0, 10)
+	existingSubs, existingTopic := n.subscriptions[topic]
+	if !existingTopic {
+		existingSubs = make([]*Subscription, 0, 10)
 	}
-	subs = append(subs, sub)
-	n.subs[topic] = subs
-
-	if isNewTopic && n.isConnActive {
-		// send to chan
-		n.subChangeCh <- &subscriptionChange{
-			topic:      topic,
-			isNewTopic: isNewTopic,
+	n.subscriptions[topic] = append(existingSubs, sub)
+
+	// We add the new subscription to the subscription list optimistically, and
+	// it needs to be done this way in case of a restart after an interrupt
+	// below has been run, but after a return to this function (say we were to
+	// add the new sub at the end of this function, it would not be picked
+	// during the restart). But in case of an error subscribing, remove the sub.
+	//
+	// By the time this function is run (i.e. after an interrupt), a lock on
+	// `n.mu` has been reacquired, and modifying subscription state is safe.
+	removeSub := func() {
+		n.subscriptions[sub.topic] = slices.DeleteFunc(n.subscriptions[sub.topic], func(s *Subscription) bool {
+			return s == sub
+		})
+
+		if len(n.subscriptions[sub.topic]) <= 1 {
+			delete(n.subscriptions, sub.topic)
 		}
 	}
 
-	return sub
-}
+	if !existingTopic {
+		// If already waiting, send an interrupt to the wait function to run a
+		// listen operation. If not, connect and listen directly, returning any
+		// errors as feedback to the caller.
+		if n.isWaiting {
+			if err := n.sendInterruptAndReceiveResult(func() error { return n.listenerListen(ctx, topic) }); err != nil {
+				removeSub()
+				return nil, err
+			}
+		} else {
+			var justConnected bool
+
+			if !n.isConnected {
+				if err := n.listenerConnect(ctx, true); err != nil {
+					removeSub()
```

Comment:
I don't think this behavior aligns with how the users of the notifier are designed today. I believe they all expect that once they call `Listen(topic)`, the notifier will do its best to make sure that `topic` is listened to until the subscriber calls `Unlisten` on the subscription. This is the case regardless of what errors/reconnects/etc happen on the notifier in the interim.

What's here seems fundamentally different. The listen attempt can _fail inline_, in which case that topic will not be listened to at all unless the caller tries to `Listen()` again. Because those subscribers are just swallowing the errors as of now (due to them being designed to consume the previous API which did not surface such errors), they will just carry on operating but without a functional listener. That doesn't seem like a desirable outcome.

Reply:
I don't think it's too different in practice to how it worked before — after `Listen` has returned successfully once, then any service that calls it is considered listening, and any subsequent failures in the machinery will cause the notifier to do its best to make sure that those subscriptions stay connected.

It may be possible (although very unlikely) currently for the producer + elector to fail on that initial `Listen` connection, but I have branches for both of them that refactor them to services that can propagate errors on `Start` like I'm proposing the notifier will here. The eventual (and soon to be reached) destination will be an initial `Listen` that fails will propagate  its error, but assuming it can succeed, will become reliably error tolerant, just as things are currently. This will also allow us to remove these sort of hacking checks that are currently required on startup because nothing propagates an error:

https://github.com/riverqueue/river/blob/4bfe88a21a2a73dce03a0a35e56970f6f11ca61e/client.go#L586-L596

Reply:
Ok, I think the issue is that there's currently no retry loop on callers to `Listen()`. This was fine previously, because no matter what, the notifier would always keep track of things that were supposed to be listened to after that initial call, regardless of any errors encountered. However with this current design, the `Listen()` call can error, the notifier will then _not_ stay subscribed to that topic, and the caller will not retry.

If you're saying you have a fast follow PR that will fix these issues, then I don't think this needs to be held up from merging, but we should probably get those additional fixes in before cutting a new release just so we're not shipping a regression like that.

Reply:
Okay, I just pushed a change that puts all listens into a retry loop with exponential backoff and a cancel check.

I don't think this is the right answer longer term, but seems like a reasonable precaution in the interim before we can start returning errors from elector/producer.